File: dart/lib/src/bankid_se.dart
import 'age_verification.dart';
import 'exceptions.dart';
import 'poll_options.dart';
import 'transport.dart';

// --- Request types ---

class BankIdSeAuthRequest {
  const BankIdSeAuthRequest({
    this.ssn,
    this.ipAddress,
    this.callbackUrl,
    this.pinRequired,
    this.intent,
    this.orgNumber,
    this.requestAddress,
    this.refId,
  });

  final String? ssn;
  final String? ipAddress;
  final String? callbackUrl;
  final bool? pinRequired;
  final String? intent;
  final String? orgNumber;
  final bool? requestAddress;
  final String? refId;

  Map<String, dynamic> toJson() => {
        if (ssn != null) 'ssn': ssn,
        if (ipAddress != null) 'ipAddress': ipAddress,
        if (callbackUrl != null) 'callbackUrl': callbackUrl,
        if (pinRequired != null) 'pinRequired': pinRequired,
        if (intent != null) 'intent': intent,
        if (orgNumber != null) 'orgNumber': orgNumber,
        if (requestAddress != null) 'requestAddress': requestAddress,
        if (refId != null) 'refId': refId,
      };
}

class BankIdSePhoneAuthRequest {
  const BankIdSePhoneAuthRequest({
    required this.ssn,
    required this.callInitiator,
    this.callbackUrl,
    this.pinRequired,
    this.intent,
    this.orgNumber,
    this.requestAddress,
    this.refId,
  });

  final String ssn;
  final String callInitiator;
  final String? callbackUrl;
  final bool? pinRequired;
  final String? intent;
  final String? orgNumber;
  final bool? requestAddress;
  final String? refId;

  Map<String, dynamic> toJson() => {
        'ssn': ssn,
        'callInitiator': callInitiator,
        if (callbackUrl != null) 'callbackUrl': callbackUrl,
        if (pinRequired != null) 'pinRequired': pinRequired,
        if (intent != null) 'intent': intent,
        if (orgNumber != null) 'orgNumber': orgNumber,
        if (requestAddress != null) 'requestAddress': requestAddress,
        if (refId != null) 'refId': refId,
      };
}

class BankIdSeSignRequest {
  const BankIdSeSignRequest({
    required this.text,
    this.ssn,
    this.ipAddress,
    this.callbackUrl,
    this.pinRequired,
    this.digest,
    this.orgNumber,
    this.requestAddress,
    this.refId,
  });

  final String text;
  final String? ssn;
  final String? ipAddress;
  final String? callbackUrl;
  final bool? pinRequired;
  final String? digest;
  final String? orgNumber;
  final bool? requestAddress;
  final String? refId;

  Map<String, dynamic> toJson() => {
        'text': text,
        if (ssn != null) 'ssn': ssn,
        if (ipAddress != null) 'ipAddress': ipAddress,
        if (callbackUrl != null) 'callbackUrl': callbackUrl,
        if (pinRequired != null) 'pinRequired': pinRequired,
        if (digest != null) 'digest': digest,
        if (orgNumber != null) 'orgNumber': orgNumber,
        if (requestAddress != null) 'requestAddress': requestAddress,
        if (refId != null) 'refId': refId,
      };
}

class BankIdSePhoneSignRequest {
  const BankIdSePhoneSignRequest({
    required this.ssn,
    required this.callInitiator,
    required this.text,
    this.callbackUrl,
    this.pinRequired,
    this.digest,
    this.orgNumber,
    this.requestAddress,
    this.refId,
  });

  final String ssn;
  final String callInitiator;
  final String text;
  final String? callbackUrl;
  final bool? pinRequired;
  final String? digest;
  final String? orgNumber;
  final bool? requestAddress;
  final String? refId;

  Map<String, dynamic> toJson() => {
        'ssn': ssn,
        'callInitiator': callInitiator,
        'text': text,
        if (callbackUrl != null) 'callbackUrl': callbackUrl,
        if (pinRequired != null) 'pinRequired': pinRequired,
        if (digest != null) 'digest': digest,
        if (orgNumber != null) 'orgNumber': orgNumber,
        if (requestAddress != null) 'requestAddress': requestAddress,
        if (refId != null) 'refId': refId,
      };
}

class BankIdSeVerifyRequest {
  const BankIdSeVerifyRequest({required this.qrCode});
  final String qrCode;
  Map<String, dynamic> toJson() => {'qrCode': qrCode};
}

class BankIdSeVerifyResponse {
  const BankIdSeVerifyResponse({
    required this.ssn,
    required this.name,
    required this.givenName,
    required this.surname,
    this.age,
    this.verifiedAt,
  });

  final String ssn;
  final String name;
  final String givenName;
  final String surname;
  final int? age;
  final String? verifiedAt;

  factory BankIdSeVerifyResponse.fromJson(Map<String, dynamic> json) =>
      BankIdSeVerifyResponse(
        ssn: json['ssn'] as String,
        name: json['name'] as String,
        givenName: json['givenName'] as String,
        surname: json['surname'] as String,
        age: json['age'] as int?,
        verifiedAt: json['verifiedAt'] as String?,
      );
}

// --- Status sealed classes ---

sealed class BankIdSeStatus {
  factory BankIdSeStatus.fromJson(Map<String, dynamic> json) =>
      switch (json['status']) {
        'PENDING' => BankIdSePending.fromJson(json),
        'COMPLETED' => BankIdSeCompleted.fromJson(json),
        'FAILED' => BankIdSeFailed.fromJson(json),
        _ => throw FormatException('Unknown bankid-se status: ${json['status']}'),
      };
}

sealed class BankIdSePhoneStatus {
  factory BankIdSePhoneStatus.fromJson(Map<String, dynamic> json) =>
      switch (json['status']) {
        'PENDING' => BankIdSePendingPhone.fromJson(json),
        'COMPLETED' => BankIdSeCompleted.fromJson(json),
        'FAILED' => BankIdSeFailed.fromJson(json),
        _ => throw FormatException('Unknown bankid-se phone status: ${json['status']}'),
      };
}

final class BankIdSePending implements BankIdSeStatus {
  const BankIdSePending({
    required this.id,
    this.refId,
    this.autoStartToken,
    this.qrStartToken,
    this.qrStartSecret,
    this.hintCode,
  });

  final String id;
  final String? refId;
  final String? autoStartToken;
  final String? qrStartToken;
  final String? qrStartSecret;
  final String? hintCode;

  factory BankIdSePending.fromJson(Map<String, dynamic> json) => BankIdSePending(
        id: json['id'] as String,
        refId: json['refId'] as String?,
        autoStartToken: json['autoStartToken'] as String?,
        qrStartToken: json['qrStartToken'] as String?,
        qrStartSecret: json['qrStartSecret'] as String?,
        hintCode: json['hintCode'] as String?,
      );
}

final class BankIdSePendingPhone implements BankIdSePhoneStatus {
  const BankIdSePendingPhone({required this.id, this.refId, this.hintCode});

  final String id;
  final String? refId;
  final String? hintCode;

  factory BankIdSePendingPhone.fromJson(Map<String, dynamic> json) =>
      BankIdSePendingPhone(
        id: json['id'] as String,
        refId: json['refId'] as String?,
        hintCode: json['hintCode'] as String?,
      );
}

final class BankIdSeCompleted implements BankIdSeStatus, BankIdSePhoneStatus {
  const BankIdSeCompleted({
    required this.id,
    this.refId,
    required this.ssn,
    required this.name,
    required this.givenName,
    required this.surname,
    this.certStartDate,
    this.address,
    this.companySignatoryText,
  });

  final String id;
  final String? refId;
  final String ssn;
  final String name;
  final String givenName;
  final String surname;
  final String? certStartDate;
  final String? address;
  final String? companySignatoryText;

  factory BankIdSeCompleted.fromJson(Map<String, dynamic> json) =>
      BankIdSeCompleted(
        id: json['id'] as String,
        refId: json['refId'] as String?,
        ssn: json['ssn'] as String,
        name: json['name'] as String,
        givenName: json['givenName'] as String,
        surname: json['surname'] as String,
        certStartDate: json['certStartDate'] as String?,
        address: json['address'] as String?,
        companySignatoryText: json['companySignatoryText'] as String?,
      );
}

final class BankIdSeFailed implements BankIdSeStatus, BankIdSePhoneStatus {
  const BankIdSeFailed({required this.id, this.refId, required this.error});

  final String id;
  final String? refId;
  final String error;

  factory BankIdSeFailed.fromJson(Map<String, dynamic> json) => BankIdSeFailed(
        id: json['id'] as String,
        refId: json['refId'] as String?,
        error: json['error'] as String,
      );
}

// --- Endpoint ---

class BankIdSeEndpoint {
  BankIdSeEndpoint(this._transport);
  final Transport _transport;

  Future<BankIdSeStatus> auth(BankIdSeAuthRequest req) async =>
      BankIdSeStatus.fromJson(await _transport.post('/v3/bankid-se/auth', req.toJson()));

  Future<BankIdSePhoneStatus> phoneAuth(BankIdSePhoneAuthRequest req) async =>
      BankIdSePhoneStatus.fromJson(await _transport.post('/v3/bankid-se/phone/auth', req.toJson()));

  Future<BankIdSeStatus> sign(BankIdSeSignRequest req) async =>
      BankIdSeStatus.fromJson(await _transport.post('/v3/bankid-se/sign', req.toJson()));

  Future<BankIdSePhoneStatus> phoneSign(BankIdSePhoneSignRequest req) async =>
      BankIdSePhoneStatus.fromJson(await _transport.post('/v3/bankid-se/phone/sign', req.toJson()));

  Future<BankIdSeVerifyResponse> verify(BankIdSeVerifyRequest req) async =>
      BankIdSeVerifyResponse.fromJson(await _transport.post('/v3/bankid-se/verify', req.toJson()));

  Future<AgeVerificationStatus> ageVerification(AgeVerificationRequest req) async =>
      AgeVerificationStatus.fromJson(await _transport.post('/v3/bankid-se/age-verification', req.toJson()));

  Future<BankIdSeStatus> authStatus(String id) async =>
      BankIdSeStatus.fromJson(await _transport.get('/v3/bankid-se/auth/$id'));

  Future<BankIdSeStatus> signStatus(String id) async =>
      BankIdSeStatus.fromJson(await _transport.get('/v3/bankid-se/sign/$id'));

  Future<AgeVerificationStatus> ageVerificationStatus(String id) async =>
      AgeVerificationStatus.fromJson(await _transport.get('/v3/bankid-se/age-verification/$id'));

  Future<void> cancelAuth(String id) => _transport.delete('/v3/bankid-se/auth/$id');
  Future<void> cancelSign(String id) => _transport.delete('/v3/bankid-se/sign/$id');
  Future<void> cancelAgeVerification(String id) => _transport.delete('/v3/bankid-se/age-verification/$id');

  Future<BankIdSeStatus> waitForAuth(String id, {PollOptions opts = const PollOptions()}) =>
      _pollStatus(() => authStatus(id), opts);

  Future<BankIdSeStatus> waitForSign(String id, {PollOptions opts = const PollOptions()}) =>
      _pollStatus(() => signStatus(id), opts);

  Future<AgeVerificationStatus> waitForAgeVerification(String id, {PollOptions opts = const PollOptions()}) =>
      _pollAgeStatus(() => ageVerificationStatus(id), opts);

  Future<BankIdSeStatus> _pollStatus(Future<BankIdSeStatus> Function() fn, PollOptions opts) async {
    final deadline = DateTime.now().add(opts.timeout);
    while (true) {
      final status = await fn();
      if (status is! BankIdSePending) return status;
      if (DateTime.now().isAfter(deadline)) throw WaitException(timeout: true);
      await Future<void>.delayed(opts.interval);
    }
  }

  Future<AgeVerificationStatus> _pollAgeStatus(
      Future<AgeVerificationStatus> Function() fn, PollOptions opts) async {
    final deadline = DateTime.now().add(opts.timeout);
    while (true) {
      final status = await fn();
      if (status is! AgeVerificationPending) return status;
      if (DateTime.now().isAfter(deadline)) throw WaitException(timeout: true);
      await Future<void>.delayed(opts.interval);
    }
  }
}

File: swift/Sources/IdkollenClient/BankIdSe.swift
import Foundation

// MARK: - Requests

public struct BankIdSeAuthRequest: Encodable {
    public let ssn: String?
    public let ipAddress: String?
    public let callbackUrl: String?
    public let pinRequired: Bool?
    public let intent: String?
    public let orgNumber: String?
    public let requestAddress: Bool?
    public let refId: String?

    public init(
        ssn: String? = nil,
        ipAddress: String? = nil,
        callbackUrl: String? = nil,
        pinRequired: Bool? = nil,
        intent: String? = nil,
        orgNumber: String? = nil,
        requestAddress: Bool? = nil,
        refId: String? = nil
    ) {
        self.ssn = ssn
        self.ipAddress = ipAddress
        self.callbackUrl = callbackUrl
        self.pinRequired = pinRequired
        self.intent = intent
        self.orgNumber = orgNumber
        self.requestAddress = requestAddress
        self.refId = refId
    }
}

public struct BankIdSePhoneAuthRequest: Encodable {
    public let ssn: String
    public let callInitiator: String
    public let callbackUrl: String?
    public let pinRequired: Bool?
    public let intent: String?
    public let orgNumber: String?
    public let requestAddress: Bool?
    public let refId: String?

    public init(
        ssn: String,
        callInitiator: String,
        callbackUrl: String? = nil,
        pinRequired: Bool? = nil,
        intent: String? = nil,
        orgNumber: String? = nil,
        requestAddress: Bool? = nil,
        refId: String? = nil
    ) {
        self.ssn = ssn
        self.callInitiator = callInitiator
        self.callbackUrl = callbackUrl
        self.pinRequired = pinRequired
        self.intent = intent
        self.orgNumber = orgNumber
        self.requestAddress = requestAddress
        self.refId = refId
    }
}

public struct BankIdSeSignRequest: Encodable {
    public let text: String
    public let ssn: String?
    public let ipAddress: String?
    public let callbackUrl: String?
    public let pinRequired: Bool?
    public let digest: String?
    public let orgNumber: String?
    public let requestAddress: Bool?
    public let refId: String?

    public init(
        text: String,
        ssn: String? = nil,
        ipAddress: String? = nil,
        callbackUrl: String? = nil,
        pinRequired: Bool? = nil,
        digest: String? = nil,
        orgNumber: String? = nil,
        requestAddress: Bool? = nil,
        refId: String? = nil
    ) {
        self.text = text
        self.ssn = ssn
        self.ipAddress = ipAddress
        self.callbackUrl = callbackUrl
        self.pinRequired = pinRequired
        self.digest = digest
        self.orgNumber = orgNumber
        self.requestAddress = requestAddress
        self.refId = refId
    }
}

public struct BankIdSePhoneSignRequest: Encodable {
    public let ssn: String
    public let callInitiator: String
    public let text: String
    public let callbackUrl: String?
    public let pinRequired: Bool?
    public let digest: String?
    public let orgNumber: String?
    public let requestAddress: Bool?
    public let refId: String?

    public init(
        ssn: String,
        callInitiator: String,
        text: String,
        callbackUrl: String? = nil,
        pinRequired: Bool? = nil,
        digest: String? = nil,
        orgNumber: String? = nil,
        requestAddress: Bool? = nil,
        refId: String? = nil
    ) {
        self.ssn = ssn
        self.callInitiator = callInitiator
        self.text = text
        self.callbackUrl = callbackUrl
        self.pinRequired = pinRequired
        self.digest = digest
        self.orgNumber = orgNumber
        self.requestAddress = requestAddress
        self.refId = refId
    }
}

public struct BankIdSeVerifyRequest: Encodable {
    public let qrCode: String
    public init(qrCode: String) { self.qrCode = qrCode }
}

public struct BankIdSeVerifyResponse: Decodable {
    public let ssn: String
    public let name: String
    public let givenName: String
    public let surname: String
    public let age: Int?
    public let verifiedAt: String?
}

// MARK: - Status enums + variants

public enum BankIdSeStatus: Decodable {
    case pending(BankIdSePending)
    case completed(BankIdSeCompleted)
    case failed(BankIdSeFailed)

    private enum DiscriminatorKey: String, CodingKey { case status }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: DiscriminatorKey.self)
        let status = try container.decode(String.self, forKey: .status)
        switch status {
        case "PENDING":   self = .pending(try BankIdSePending(from: decoder))
        case "COMPLETED": self = .completed(try BankIdSeCompleted(from: decoder))
        case "FAILED":    self = .failed(try BankIdSeFailed(from: decoder))
        default:
            throw DecodingError.dataCorruptedError(forKey: .status, in: container, debugDescription: "Unknown bankid-se status: \(status)")
        }
    }
}

public enum BankIdSePhoneStatus: Decodable {
    case pending(BankIdSePendingPhone)
    case completed(BankIdSeCompleted)
    case failed(BankIdSeFailed)

    private enum DiscriminatorKey: String, CodingKey { case status }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: DiscriminatorKey.self)
        let status = try container.decode(String.self, forKey: .status)
        switch status {
        case "PENDING":   self = .pending(try BankIdSePendingPhone(from: decoder))
        case "COMPLETED": self = .completed(try BankIdSeCompleted(from: decoder))
        case "FAILED":    self = .failed(try BankIdSeFailed(from: decoder))
        default:
            throw DecodingError.dataCorruptedError(forKey: .status, in: container, debugDescription: "Unknown bankid-se phone status: \(status)")
        }
    }
}

public struct BankIdSePending: Decodable {
    public let id: String
    public let refId: String?
    public let autoStartToken: String?
    public let qrStartToken: String?
    public let qrStartSecret: String?
    public let hintCode: String?
}

public struct BankIdSePendingPhone: Decodable {
    public let id: String
    public let refId: String?
    public let hintCode: String?
}

public struct BankIdSeCompleted: Decodable {
    public let id: String
    public let refId: String?
    public let ssn: String
    public let name: String
    public let givenName: String
    public let surname: String
    public let certStartDate: String?
    public let address: String?
    public let companySignatoryText: String?
}

public struct BankIdSeFailed: Decodable {
    public let id: String
    public let refId: String?
    public let error: String
}

// MARK: - Endpoint

public final class BankIdSeEndpoint {
    private let transport: Transport
    internal init(_ transport: Transport) { self.transport = transport }

    public func auth(_ req: BankIdSeAuthRequest) async throws -> BankIdSeStatus {
        try await transport.post("/v3/bankid-se/auth", body: req)
    }

    public func phoneAuth(_ req: BankIdSePhoneAuthRequest) async throws -> BankIdSePhoneStatus {
        try await transport.post("/v3/bankid-se/phone/auth", body: req)
    }

    public func sign(_ req: BankIdSeSignRequest) async throws -> BankIdSeStatus {
        try await transport.post("/v3/bankid-se/sign", body: req)
    }

    public func phoneSign(_ req: BankIdSePhoneSignRequest) async throws -> BankIdSePhoneStatus {
        try await transport.post("/v3/bankid-se/phone/sign", body: req)
    }

    public func verify(_ req: BankIdSeVerifyRequest) async throws -> BankIdSeVerifyResponse {
        try await transport.post("/v3/bankid-se/verify", body: req)
    }

    public func ageVerification(_ req: AgeVerificationRequest) async throws -> AgeVerificationStatus {
        try await transport.post("/v3/bankid-se/age-verification", body: req)
    }

    public func authStatus(id: String) async throws -> BankIdSeStatus {
        try await transport.get("/v3/bankid-se/auth/\(id)")
    }

    public func signStatus(id: String) async throws -> BankIdSeStatus {
        try await transport.get("/v3/bankid-se/sign/\(id)")
    }

    public func ageVerificationStatus(id: String) async throws -> AgeVerificationStatus {
        try await transport.get("/v3/bankid-se/age-verification/\(id)")
    }

    public func cancelAuth(id: String) async throws {
        try await transport.delete("/v3/bankid-se/auth/\(id)")
    }

    public func cancelSign(id: String) async throws {
        try await transport.delete("/v3/bankid-se/sign/\(id)")
    }

    public func cancelAgeVerification(id: String) async throws {
        try await transport.delete("/v3/bankid-se/age-verification/\(id)")
    }

    public func waitForAuth(id: String, opts: PollOptions = .init()) async throws -> BankIdSeStatus {
        let deadline = Date().addingTimeInterval(opts.timeout)
        while true {
            let status = try await authStatus(id: id)
            if case .pending = status {
                if Date() >= deadline { throw WaitError(timeout: true) }
                try await Task.sleep(nanoseconds: UInt64(opts.interval * 1_000_000_000))
            } else {
                return status
            }
        }
    }

    public func waitForSign(id: String, opts: PollOptions = .init()) async throws -> BankIdSeStatus {
        let deadline = Date().addingTimeInterval(opts.timeout)
        while true {
            let status = try await signStatus(id: id)
            if case .pending = status {
                if Date() >= deadline { throw WaitError(timeout: true) }
                try await Task.sleep(nanoseconds: UInt64(opts.interval * 1_000_000_000))
            } else {
                return status
            }
        }
    }

    public func waitForAgeVerification(id: String, opts: PollOptions = .init()) async throws -> AgeVerificationStatus {
        let deadline = Date().addingTimeInterval(opts.timeout)
        while true {
            let status = try await ageVerificationStatus(id: id)
            if case .pending = status {
                if Date() >= deadline { throw WaitError(timeout: true) }
                try await Task.sleep(nanoseconds: UInt64(opts.interval * 1_000_000_000))
            } else {
                return status
            }
        }
    }
}
